src/compiler/scala/tools/reflect/StringInterpolator.scala

```diff
@@ -0,0 +1,162 @@
+    body
+    true
+  }
+
+  @inline private def falsely(body: => Unit): Boolean = {
+    body
+    false
+  }
+
+  private def bail(msg: String) = global.abort(msg)
+
+  def interpolateString: Tree = interpolateImpl(false)
+
+  def interpolateRaw: Tree = interpolateImpl(true)
+
+  def interpolateImpl(raw: Boolean): Tree = c.macroApplication match {
+    case Applied(Select(Apply(_, parts), _), _, argss) =>
+      val args = argss.flatten
+
+      def badlyInvoked = (parts.length != args.length + 1) && truly {
+        //probably cant use s"..." here as this is the macro for s
+        def because(s: String) = "too " + s + " arguments for interpolated string"
+
+        val (p, msg) =
+          if (args.length + 1 < parts.length)
+            (if (args.isEmpty) c.enclosingPosition else args.last.pos, because("few"))
+          else (args(parts.length - 1).pos, because("many"))
+        c.abort(p, msg)
+      }
+
+      if (badlyInvoked) c.macroApplication else {
+        val res = interpolated(raw, parts, args)
+        /*if (settings.debug) */reporter.echo(c.enclosingPosition, "rewrite expression to "+res)
+        res
+      }
+    case other =>
+      bail("Unexpected application " + showRaw(other))
+      other
+  }
+
+  def interpolated(raw: Boolean, parts: List[Tree], args: List[Tree]): Tree = {
+    var literalSize = 0
+    val literals: List[List[Literal]] = parts map {
+      case lit@Literal(Constant(x: String)) =>
+        if (x.isEmpty) Nil
+        else {
+          literalSize += x.length
+          if (raw) lit :: Nil
+          else (Literal(Constant(/*StringContext.*/ treatEscapes(x))) setPos (lit.pos)) :: Nil
+        }
+      case _ => throw new IllegalArgumentException("internal error: argument parts must be a list of string literals")
+    }
+    if (args isEmpty)
+    //it of the form s"some text" or s""
+      parts(0)
+    else if (args.tail.isEmpty && literalSize == 0)
+    //it of the form s"$value", so we dont need a StringBuilder
+      Apply(Select(Ident(StringModule), termNames.valueOf), args.head :: Nil) setPos (args.head.pos)
+    else {
+
+      //if we know for the literals that there is a large text block, then we may as well create it with an appropriate size
+      val cons = if (literalSize > 16) Literal(Constant(java.lang.Integer.highestOneBit(literalSize) << 1)) :: Nil else Nil
+      var expr = Apply(Select(New(Ident(JavaStringBuilderClass)), termNames.CONSTRUCTOR), cons)
+
+      val append = TermName("append")
+
+      val literalsIt = literals.iterator
+      val argsIt = args.iterator
+      val first = literalsIt.next()
+      if (!first.isEmpty) expr = Apply(Select(expr, append), first) setPos (first.head.pos)
+      while (literalsIt.hasNext) {
+        val literal = literalsIt.next()
+        val arg = argsIt.next()
+        expr = Apply(Select(expr, append), arg :: Nil) setPos (arg.pos)
+        if (!literal.isEmpty) expr = Apply(Select(expr, append), literal) setPos (literal.head.pos)
+      }
```

Comment:
Similarly, this also changes semantics:

```
scala> s"foo${new Array[Char](0)}bar"
res17: String = foo[C@7b41a3f9bar

scala> sNew"foo${new Array[Char](0)}bar"
<console>:37: rewrite expression to new StringBuilder().append("foo").append(new Array[Char](0)).append("bar").toString()
       sNew"foo${new Array[Char](0)}bar"
       ^
res18: String = foobar
```

If `expr.tpe` is a numeric value class, I believe it is safe to call the corresponding primitive overload of `apppend`. But if it an `Array` / `CharSequence` / `StringBuffer`, we need to avoid the overloads.